### ISSUE TITLE
Make non-sensitive portions of the trace API use accessible to all users

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#18734]
Since the problem keeps popping up from time to time and there seems to be no FR for it, I'm opening one here.

As discussed at various places including https://github.com/openssl/openssl/issues/18691#issuecomment-1176053044, it is a pity that regular OpenSSL users (those that do not compile it themselves) cannot make use of the trace API output.

The reason not to enable it statically was that it may reveal sensitive internal information.
While this certainly holds for some aspects, for the majority of uses this is likely not the case.
So I suggest supporting by default to enable those trace categories that do not output sensitive information.

To be more precise, let's have a whitelist of categories for which tracing can be enabled by anyone at runtime.
This will then include CMP, in the future X509_VERIFY (which I have in the pipeline), and likely a couple of more.

**Comment by levitte** [comment]
There isn't really a distinction between sensitive and non-sensitive data in the trace API.  Whatever the printing calls decide to output is what gets output.

However, @mattcaswell floated the idea of syslog-like levels (INFO, DEBUG, etc), and we might want to use something like that to make the distinction.

**Comment by levitte** [comment]
Also, a lot of the team is on vacation for the moment, so engaged discussions will probably not happen before everyone is back.

**Comment by DDvO** [comment]
> There isn't really a distinction between sensitive and non-sensitive data in the trace API. Whatever the printing calls decide to output is what gets output.

Yes, but for each category someone will know whether anything sensitive (such as private/symmetric key material) can get printed, and if not, can put this category on the whitelist.

> However, @mattcaswell floated the idea of syslog-like levels (INFO, DEBUG, etc), and we might want to use something like that to make the distinction.

The severity is actually a different dimension. Would be nice to add this, too.

You may recall that for the CMP implementation I came up with a workaround for the inaccessibility (by default) of the trace output in `crypto/cmp/cmp_util.c` including
```
int OSSL_CMP_print_to_bio(BIO *bio, const char *component, const char *file,
                          int line, OSSL_CMP_severity level, const char *msg)
```
This facility includes the syslog-like level of severity. Users can decide which level of verbosity they want,
and if a message is output because it is above the threshold, it includes an indication of its severity
(EMERG, ALERT, CRIT, ERR, WARN, NOTE, INFO, and DEBUG, and this can be extended further down to include TRACE).

> Also, a lot of the team is on vacation for the moment, so engaged discussions will probably not happen before everyone is back.

Good to know.

**Comment by DDvO** [comment]
Any logging feature that supports the severity level TRACE could use the existing trace API for output at level TRACE.

I found that the logging functionality in `cmp_util.c` could be easily be extended to include the severity level TRACE and generalized (and simplified) for use with any other part of OpenSSL.

**Comment by DDvO** [comment]
Due to a recent fix: [TRACE: automatically respect disabled categories (#25652)](https://github.com/openssl/openssl/pull/25652#top)
it can no more happen that trace out happens without being explicitly enabled both statically (at build time) and dynamically (via the `OPENSSL_TRACE` environment variable).

Can't we now statically enable trace support by default?
This is still safeguarded via `OPENSSL_TRACE`.

**Comment by t8m** [comment]
**OTC: Should we enable-trace by default?**

**Comment by t8m** [comment]
OTC discussed this. We are concerned about the overhead and potential security implications if applications misuse the trace API. We would be OK with enabling trace in debug builds by default.

**Comment by DDvO** [comment]
> OTC discussed this. 

Thank you for re-considering this in the new improved situation 
where trace output (likely) no more can happen by mistake even if enabled statically.

> We are concerned about the overhead

In which sense would this be an overhead?
Do you mean code size (is this significant in comparison to the overall code size)
and/or the execution time (where calls could be in-lined) for checking dynamic enabledness?
Anyway, tracing hopefully is not used in tight loops (where it likely would produce too much data).

> and potential security implications if applications misuse the trace API.

Isn't it always in the responsibility of applications which data they output to where?

> We would be OK with enabling trace in debug builds by default.

While this would just be a minor convenience improvement (because debug builds anyway require adding the `--debug` config option, where adding in addition ` enable-trace` is just a minor extra thing), this would still already be a nice step, 
because it would at least remind developers that the trace API exists and is useful
and may lead to better immediate test coverage.

Yet for regular users the trace API is still not feasible to use unless enabled also for productive builds.
This was the reason for me to add to the CMP implementation [an extra mechanism](https://github.com/openssl/openssl/blob/012353bdf21b98def920ac317b94c4a9ed501b79/crypto/cmp/cmp_local.h#L837) for providing detail in particular about cert validation issues.
